### ISSUE TITLE
Dafny 4.0 suggestion: Get rids of `returns` keyword to unify function, predicate and methods

### ISSUE THREAD

**Issue by MikaelMayer** [dafny-lang/dafny#2342]
We could unify the returns syntax for both methods and functions.
This would make it easier to convert functions to methods and vice-versa, lower the cognitive burden.
I find myself several times writing `returns` at the wrong place or omitting it sometimes. Let's fix that.

Instead of
```
predicate p(x: int) ensures p(x) ==> x > 0 {  }
function f(x: int): (z: int) {  }
method g(x: int) returns (z: int, w: int) {  }
method h(x: int) returns (z: int) {
  if x > 0 {
    z := x + 1;
    return; 
  }
}
```

we would write
```
predicate p(x: int): (z: bool) ensures z ==> x > 0 {  } (ability to name predicate's output)
function f(x: int): (z: int) {  } (unchanged)
method g(x: int): (z: int, w: int): int {  }  // Breaking change
method h(x: int): int { // Allow unamed output
  if x > 0 {
    return x + 1; // The alternative z := x + 1 is not possible anymore
  }
}
```

To reason about the output, the method should always have a name. But for methods inside classes that routinely modify fields, the return value should have the right to not have a name if it's not relevant for the specification. Like for functions, we can add the name of the output later.

**Comment by MikaelMayer** [comment]
Hmmm that's a very good point.

I think it's the time to drop the parentheses of the return variable name. It will make it easier to add a name to the result (no need to add a closing parenthesis !)
```dafny
function f(x: int): int
function f(x: int): result: int
method wat(x: int): int, int
method wat(x: int): a: int, b: int
```
Plus, it will remove the existing ambiguity that it looks like methods return a tuple of values, when they actually return two values.

Alternatively, if we want to avoid the ambiguity of the colon, we could simply use another syntax:
```dafny
function f(x: int) -> int
function f(x: int) -> result: int
method wat(x: int) -> int, int
method wat(x: int) -> a: int, b: int
```
Let's be fancy for this one, which mimicks in reverse the assignment `var a: int, b: int := wat(2);`:
```dafny
function f(x: int) =: int
function f(x: int) =: result: int
method wat(x: int) =: int, int
method wat(x: int) =: a: int, b: int
```

Another alternative could consist of always requiring `returns` and parentheses
```dafny
function f(x: int) returns (int)
function f(x: int) returns (result: int)
method wat(x: int) returns (int, int)
method wat(x: int) returns (a: int, b: int)
```

**Comment by robin-aws** [comment]
```dafny
method g(x: int): (z: int, w: int): int {  }  // Breaking change
```

Is that final `: int` a typo?

**Comment by robin-aws** [comment]
As a minimal improvement, I like the idea of allowing either `:` or `returns` between the parameter list and the result list, which would not be a breaking change.

**Comment by MikaelMayer** [comment]
I like this idea too. That way, one could migrate from `: int` to `: (name: int)` to `returns (name: int, name2: int)` as needed.

**Comment by kjx** [comment]
Yeah, unify this. Causes lots of problems for students learning dafny. for anyone who needs to move code between methods and functions, is a gratuitous inconsistency, and pretty much indefensible.

I like the "->" syntax, of course, because that's what we picked for Grace. 

I could argue ":" is misleading there because the type of the whole function its some arrow type,
not just the return type.

**Comment by robin-aws** [comment]
At the risk of scope creep, it's also worth considering unifying expression and statement syntax if possible, so switching between `function` and `method` is literally all that's necessary:

```dafny
function h(x: int): int {
  if x > 0 {
    return x + 1;
  }
}
```

**Comment by kjx** [comment]
well I support unifying the syntax, for what it's worth - although I'd keep "then".

probably should to move "let"s, to, well "lets"...

**Comment by robin-aws** [comment]
(Hit enter before I was done with that comment above :)

I'm actually more inclined to keep the `var x := ...; ...` syntax since that's already semi-aligned between expressions and statements. I'd rather the parser allow all the same features, and then have the resolver reject things that can't be used in a function/expression, such as `x := y;`
